Source/santad/EventProviders/FAAPolicyProcessor.mm

```diff
@@ -121,30 +251,244 @@
   return [decision_cache_ cachedDecisionForFile:stat_buf];
 }
 
-static inline std::string Path(const es_file_t *esFile) {
-  return std::string(esFile->path.data, esFile->path.length);
+bool FAAPolicyProcessor::PolicyAllowsReadsForTarget(
+    const Message &msg, const PathTarget &target,
+    const std::shared_ptr<WatchItemPolicyBase> policy) {
+  // All special cases currently require the option "allow_read_access" is set
+  if (!policy->allow_read_access) {
+    return false;
+  }
+
+  switch (msg->event_type) {
+    case ES_EVENT_TYPE_AUTH_OPEN:
+      // If the policy is write-only, but the operation isn't a write action, it's allowed
+      if (!(msg->event.open.fflag & kOpenFlagsIndicatingWrite)) {
+        return true;
+      }
+      break;
+
+    case ES_EVENT_TYPE_AUTH_CLONE:
+      // If policy is write-only, readable targets are allowed (e.g. source file)
+      if (target.is_readable) {
+        return true;
+      }
+      break;
+
+    case ES_EVENT_TYPE_AUTH_COPYFILE:
+      // Note: Flags for the copyfile event represent the kernel view, not the usersapce
+      // copyfile(3) implementation. This means if a `copyfile(3)` flag like `COPYFILE_MOVE`
+      // is specified, it will come as a separate `unlink(2)` event, not a flag here.
+      if (target.is_readable) {
+        return true;
+      }
+      break;
+
+    case ES_EVENT_TYPE_AUTH_CREATE:
+    case ES_EVENT_TYPE_AUTH_EXCHANGEDATA:
+    case ES_EVENT_TYPE_AUTH_LINK:
+    case ES_EVENT_TYPE_AUTH_RENAME:
+    case ES_EVENT_TYPE_AUTH_TRUNCATE:
+    case ES_EVENT_TYPE_AUTH_UNLINK:
+    default:
+      // These event types have no special case
+      break;
+  }
+
+  return false;
 }
 
-static inline std::string Path(const es_string_token_t &tok) {
-  return std::string(tok.data, tok.length);
+FileAccessPolicyDecision FAAPolicyProcessor::ApplyPolicy(
+    const Message &msg, const PathTarget &target,
+    const std::optional<std::shared_ptr<WatchItemPolicyBase>> optional_policy,
+    CheckIfPolicyMatchesBlock check_if_policy_matches_block) {
+  if (!optional_policy.has_value()) {
+    return FileAccessPolicyDecision::kNoPolicy;
+  }
+
+  // If the process is signed but has an invalid signature, it is denied
+  if (((msg->process->codesigning_flags & (CS_SIGNED | CS_VALID)) == CS_SIGNED) &&
+      [configurator_ enableBadSignatureProtection]) {
+    // TODO(mlw): Think about how to make stronger guarantees here to handle
+    // programs becoming invalid after first being granted access. Maybe we
+    // should only allow things that have hardened runtime flags set?
```

Comment:
I think if BadSignatureProtection is turned on we _should_ do this **or** add a field in the policy that enables this

Reply:
Maybe, but I think it's still opt-in for a lot of things which would make this difficult to enforce and let people keep software they use. Should definitely chat more about this though.